### PR TITLE
Allow passing filename for documents

### PR DESCRIPTION
**Pull request by samuelhwilliams**
At the moment, when files are downloaded they simply have a UUID and file extension. This isn't really useful for users as it won't ever help them recognise what the file is.

Let's allow services to pass filenames through, so that when the file is downloaded it has a service-customisable filename.

We will now send a 'Content-Disposition' header for all requests to the 'download document' endpoint. We force it as an attachment (which triggers download in most browsers) for the same set of file extensions as before - csv and rtf. For all others, we mark it as 'inline', telling the browser to open it in a tab. But by specifying the content disposition, we can include a filename that will auto-populate file dialogs with the correct default filename.

If `filename` is passed in the data for uploading a new document, this overrides any `is_csv` parameter. We may mark `is_csv` as deprecated and prefer the `filename` data option - but the documentation needs to be clear on this behaviour either way.

---

🚨⚠️ This will be deployed automatically all the way to production when you click merge ⚠️🚨

For more information, including how to check this deployment on preview or staging first before it goes to production, see our [team wiki section on deployment](https://github.com/alphagov/notifications-manuals/wiki/Merging-and-deploying#deployment)

### CODE REVIEW

File: app/utils/validation.py [hunk 1]
```diff
@@ -33,3 +36,15 @@ def clean_and_validate_retention_period(retention_period):
         retention_period += "s"
 
     return retention_period
+
+
+def clean_and_validate_filename(filename):
+    if "." not in filename:
+        raise ValueError("Filename must have a file extension, eg .csv")
```
Comment (karlchillmaid) [inline app/utils/validation.py:43]:
```suggestion
        raise ValueError("The file name must end with a file extension. For example, filename.csv")
```

@samuelhwilliams let’s discuss this.

**Review by CrystalPea — COMMENTED** [review]
Nice work, I think one test may be missing.

File: app/utils/files.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from typing import Tuple
+
+
+def split_filename(filename: str, *, dotted: bool) -> Tuple[str, str]:
+    *parts, ext = filename.split(".")
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/utils/files.py:5]: Yes, a filename could be `my.silly.dotted.file.txt`, so parts in that case would be `['my', 'silly', 'dotted', 'file']` and ext would be `txt`.

File: app/utils/validation.py (hunk 1 above)
Comment (samuelhwilliams) [reply app/utils/validation.py:43]:
```suggestion
        raise ValueError("`filename` must end with a file extension. For example, filename.csv")
```

File: tests/utils/test_validation.py [hunk 3]
```diff
@@ -27,3 +28,21 @@ def test_clean_and_validate_retention_period_invalid_values(value):
         clean_and_validate_retention_period(value)
 
     assert str(e.value) == "Retention period must be a string of the format '<1-78> weeks'."
+
+
+def test_clean_and_validate_filename_needs_dot():
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply tests/utils/test_validation.py:38]: done :)

File: tests/utils/test_validation.py [hunk 4]
```diff
@@ -27,3 +28,26 @@ def test_clean_and_validate_retention_period_invalid_values(value):
         clean_and_validate_retention_period(value)
 
     assert str(e.value) == "Retention period must be a string of the format '<1-78> weeks'."
+
+
+@pytest.mark.parametrize("filename", ("file.csv", "my-file.csv", "my.dotted.file.csv", "!@£$%^&*().pdf"))
+def test_clean_and_validate_filename_happy_path(client, filename):
+    clean_and_validate_filename(filename)
```
Comment (CrystalPea) [inline tests/utils/test_validation.py:35]: don't we need an assertion?

**Review by CrystalPea — APPROVED** [review]

**Pull request merged**